### ISSUE TITLE
http1 overperforms h2c

### ISSUE THREAD

**Issue by kptfh** [eclipse/jetty.project#3206]
Running reactive feign over h2c transport was expecting to get better performance. But benchmark shows significant decrease (50%). Is it expected behaviour?

Here is the results:
RealParallelRequestBenchmarks.feignJetty         thrpt   15  162.061 ± 61.232  ops/s
RealParallelRequestBenchmarks.feignJettyH2c      thrpt   15  105.561 ± 55.021  ops/s

You can run it on your own (just run `main` method)
https://github.com/Playtika/feign-reactive/blob/feature/jetty-http2/feign-reactor-benchmarks/src/main/java/reactivefeign/benchmarks/RealParallelRequestBenchmarks.java

**Comment by gregw** [comment]
@kptfh can you tell us a bit more about what these "real parallel request benchmarks" are doing and what sort of network you are running them over?

h2 can be more performant, but it is not necessarily so.  h2 tries to get more performance out of a TCP/IP connection by warming up it's TCP/IP slow start window and then using that for many requests, but if you compare that to h1 which might open many connections, then it can get many small slow start windows and eventually a larger combined window.... which can work fine, unless you are fighting against many other clients all trying to out compete by creating many connections.

h2 can be more peformant because of header compression.... but only if bandwidth constrained rather than CPU constrained.  So benchmarks conducted over fast local networks may not see that.

h2 also has it's own flow control algorithms, which may need to be tuned and correctly sized for your traffic.

h2 is also concerned with time to first display a page, so it looks at latency from idle to having served all the required resources for a page.  Such low latency rendering does not automatically mean better throughput.  It may be that few users get much better quality of service.

So h2 or h2c are not automatic wins for throughput, specially on non-realworld benchmarks.   In fact the extra work that we have to do to support h2c in terms of compression, encoding, multiplexing, can definitely consume more server resources... specially of the server is being hammered by a fast network.

**Comment by kptfh** [comment]
@gregw Thank you a lot for the comprehensive answer. 
This benchmark starts local server and sends many h1 and h2c requests in parallel.
Now it's clear that this synthetic benchmark is very biased and doesn't represent real cases.

Have just 2 follow up  questions:
1) We plan to use this feign client for cross microservices communication.
    From above mentioned reason I can assume that we don't get any performance improvements as:
    - fast network
    - relatively small number of clients (only other microservices) 
    What is your filling about it?
2) In low level `HTTP2Client` we setup TCP session and then use it obviously to send requests. 
    How does it managed in high level `HttpClient` over http2 transport? Does it cache TCP sessions internally? When I run many request in parallel will HttpClient tunnel them in one TCP connection?

**Comment by kptfh** [comment]
Sometimes while running benchmark I get 
`java.io.IOException: 1/unexpected_rst_stream_frame`
Should I open new issue?

**Comment by sbordet** [comment]
@kptfh yes please, open a new issue for the `1/unexpected_rst_stream_frame`.

**Comment by sbordet** [comment]
@kptfh HTTP/2 should be than HTTP/1.1 when you have a large number of requests/s from one node to another, as less connections will be opened.

When `HttpClient` is used with the HTTP/2 transport, it will pool connections (as it does for the HTTP/1.1 transport), it will queue requests if it cannot immediately send them (e.g. connection not open yet) and yes it will multiplex requests onto a connection until `maxConcurrentStream` is reached (then it will open a new connection).

**Comment by sbordet** [comment]
I think this issue can be closed as it has been answered.
@kptfh let us know if you have further comments about this issue.
